### PR TITLE
Async promo dialog

### PR DESCRIPTION
**Pull request by aforcier**
Adds a promo dialog, shown the first time a user presses the 'Publish' button:

async-promo-dialog-pr

**Note:** For now, the 'How does it work?' text is hidden. Closer (or just after) the beta we'll put up a post on mobile make, link to that, and unhide the link text.

I also added a new `PromoDialogAdvanced`, intended for reuse for any dialogs requiring a link or negative button field (the existing `PromoDialogEditor` now extends this one, adding an extra title field for 'Beta'). The pre-existing `PromoDialog` can be used for simple (image, title, description, positive button) promos.

Also some design changes:
* All promo dialogs (currently this includes Aztec, the stats widget, and this one) now have a divider above the button row:

dialogs-with-divider

cc @iamthomasbishop (no visual changes to the async promo since we last spoke)

### CODE REVIEW

File: WordPress/src/main/java/org/wordpress/android/modules/AppComponent.java [hunk 1]
```diff
@@ -26,6 +25,7 @@
 import org.wordpress.android.ui.accounts.login.Login2FaFragment;
 import org.wordpress.android.ui.accounts.login.LoginEmailFragment;
 import org.wordpress.android.ui.accounts.login.LoginEmailPasswordFragment;
+import org.wordpress.android.ui.accounts.login.LoginEpilogueFragment;
```
Comment (mzorz) [inline WordPress/src/main/java/org/wordpress/android/modules/AppComponent.java]: nice name

File: WordPress/src/main/java/org/wordpress/android/modules/AppComponent.java [hunk 2]
```diff
@@ -221,7 +221,7 @@
     void inject(ReaderWebView object);
     void inject(ReaderPostPagerActivity object);
 
-    void inject(EditorReleaseNotesActivity object);
+    void inject(ReleaseNotesActivity object);
```
Comment (mzorz) [inline WordPress/src/main/java/org/wordpress/android/modules/AppComponent.java]: nice reuse

File: WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java [hunk 3]
```diff
@@ -471,11 +473,17 @@ public void onConfigurationChanged(Configuration newConfig) {
         if (orientation != mPhotoPickerOrientation) {
             resizePhotoPicker();
         }
+
+        // If we're showing the Async promo dialog, we need to redraw it so it takes into account
```
Comment (mzorz) [inline WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java]: not sure what the comment meant, it should continue but ends abruptly?

File: WordPress/src/main/java/org/wordpress/android/ui/posts/PromoDialogAdvanced.java [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+import org.wordpress.android.R;
+import org.wordpress.android.util.DisplayUtils;
+import org.wordpress.android.widgets.WPTextView;
+
+/**
+ * Similar to {@link PromoDialog}, but with an optional link field and negative button.
+ */
+public class PromoDialogAdvanced extends PromoDialog {
+    public static class Builder extends PromoDialog.Builder {
+        @StringRes int linkId;
```
Comment (mzorz) [inline WordPress/src/main/java/org/wordpress/android/ui/posts/PromoDialogAdvanced.java]: can we make these `linkResId`,  `buttonNegativeResId`, `descriptionResId` etc?

File: WordPress/src/main/java/org/wordpress/android/ui/posts/PromoDialogAdvanced.java [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+    protected int mButtonNegativeId;
+    protected int mLinkId;
+
+    protected View.OnClickListener mNegativeButtonOnClickListener;
+    protected View.OnClickListener mLinkOnClickListener;
+
+    protected static PromoDialogAdvanced newInstance(Builder builder) {
+        PromoDialogAdvanced fragment = new PromoDialogAdvanced();
+        Bundle args = new Bundle();
+        args.putInt("drawableId", builder.drawableId);
```
Comment (mzorz) [inline WordPress/src/main/java/org/wordpress/android/ui/posts/PromoDialogAdvanced.java]: Can we move the "drawableId", "titleId", etc strings to `private static final String...`?

File: WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java (hunk 3 above)
Comment (aforcier) [inline WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java]: Woops! Fixed in 0a81c29.

File: WordPress/src/main/java/org/wordpress/android/ui/posts/PromoDialogAdvanced.java (hunk 4 above)
Comment (aforcier) [inline WordPress/src/main/java/org/wordpress/android/ui/posts/PromoDialogAdvanced.java]: Done in f4e7063.

File: WordPress/src/main/java/org/wordpress/android/ui/posts/PromoDialogAdvanced.java (hunk 5 above)
Comment (aforcier) [inline WordPress/src/main/java/org/wordpress/android/ui/posts/PromoDialogAdvanced.java]: An excellent idea - done in f4e7063.

**Comment by iamthomasbishop** [comment]
This is looking great! I'm thinking the illustration should have slightly less top/bottom padding, but let's roll with this for now.

File: WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java [hunk 6]
```diff
@@ -2507,6 +2519,27 @@ public void onClick(View view) {
...
     }
 
+    private void showAsyncPromoDialog() {
+        PromoDialogAdvanced asyncPromoDialog = new PromoDialogAdvanced.Builder(
+                R.drawable.img_promo_async,
+                R.string.async_promo_title,
+                R.string.async_promo_description,
+                android.R.string.ok)
+                // TODO: Re-enable once a release notes page exists for Async
+//                .setLinkText(R.string.async_promo_link)
```
Comment (mzorz) [inline WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java]: let's create an issue so we track this one before release

File: WordPress/src/main/java/org/wordpress/android/ui/posts/PromoDialog.java [hunk 7]
```diff
@@ -53,19 +79,33 @@ public View onCreateView(LayoutInflater inflater, ViewGroup container,
...
+        btn.setText(mButtonPositiveId);
+        if (mPositiveButtonOnClickListener == null) {
+            btn.setOnClickListener(new View.OnClickListener() {
+                @Override
+                public void onClick(View v) {
+                    getDialog().cancel();
+                }
+            });
+        } else {
+            btn.setOnClickListener(mPositiveButtonOnClickListener);
```
Comment (mzorz) [inline WordPress/src/main/java/org/wordpress/android/ui/posts/PromoDialog.java]: We're lacking a dialog `cancel()` call here like we do in https://github.com/wordpress-mobile/WordPress-Android/pull/6467/files#diff-1999fe3664a86fade02c8922f6857c75R132 

The default click handler does issue a cancel:
https://github.com/wordpress-mobile/WordPress-Android/pull/6467/files#diff-42c777438a1a186bb67bd763a279089eR91
I’m wondering, shouldn’t we make sure that when you set a handler we _also_  cancel the dialog, either by giving the new handler that responsibility, or if not, we could maybe set our own anonymous handler and call the other listener within it, like this:
```
            btn.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {
                    getDialog().cancel();
                    mPositiveButtonOnClickListener.onClick(view);
                }
            });
```

File: WordPress/src/main/java/org/wordpress/android/ui/posts/PromoDialogAdvanced.java [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+        if (mButtonNegativeId != 0) {
+            buttonNegative.setText(mButtonNegativeId);
+            buttonNegative.setOnClickListener(mNegativeButtonOnClickListener);
+        } else {
+            buttonNegative.setVisibility(View.GONE);
+        }
+
+        Button buttonPositive = (Button) view.findViewById(R.id.promo_dialog_button_positive);
+        buttonPositive.setText(mButtonPositiveId);
+        buttonPositive.setOnClickListener(mPositiveButtonOnClickListener);
```
Comment (mzorz) [inline WordPress/src/main/java/org/wordpress/android/ui/posts/PromoDialogAdvanced.java]: None of this is necessary as the positive button's text and listener are already being set in the base class

File: WordPress/src/main/java/org/wordpress/android/ui/posts/PromoDialog.java (hunk 7 above)
Comment (aforcier) [inline WordPress/src/main/java/org/wordpress/android/ui/posts/PromoDialog.java]: I'm not sure we should hard-code that behavior. It's conceivable that you might not want to dismiss the dialog on the positive button, or do something else with the dialog first, so to me it makes sense to leave it up to the dialog 'client' to dictate the behavior.

File: WordPress/src/main/java/org/wordpress/android/ui/posts/PromoDialogAdvanced.java (hunk 8 above)
Comment (aforcier) [inline WordPress/src/main/java/org/wordpress/android/ui/posts/PromoDialogAdvanced.java]: I agree it's not necessary as things currently stand. But, we don't have a way of checking that the layout we're using matches the layout of the parent class. So I left in the setup so it's explicit what view names each dialog expects to be there. This is around where the usefulness of the inheritance pattern wears thin 😞

File: WordPress/src/main/java/org/wordpress/android/ui/posts/PromoDialog.java (hunk 7 above)
Comment (mzorz) [inline WordPress/src/main/java/org/wordpress/android/ui/posts/PromoDialog.java]: 👍 ok

File: WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java (hunk 6 above)
Comment (aforcier) [inline WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java]: 👍 created https://github.com/wordpress-mobile/WordPress-Android/issues/6470.

File: WordPress/src/main/java/org/wordpress/android/ui/posts/PromoDialogAdvanced.java (hunk 8 above)
Comment (mzorz) [inline WordPress/src/main/java/org/wordpress/android/ui/posts/PromoDialogAdvanced.java]: Let's leave it as is - I think the layouts will  be kept similar for the context of use we are giving to these Dialogs, but really no harm in leaving this and re-inforcing the need to have a specific `promo_dialog_button positive` set in the view.

**Comment by mzorz** [comment]
Nice job @aforcier @iamthomasbishop !
Tested this on a couple of handsets and emus, even very small screens, and works alright 👍 
:shipit:

**Pull request merged**